modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRequestCollection.java

```diff
@@ -0,0 +1,343 @@
+        Map<StatsKeyMessage, CacheGroupContext> keyGroups = new HashMap<>();
+        for (Map.Entry<CacheGroupContext, Collection<StatsKeyMessage>> grpKeys : grpContexts.entrySet())
+            for(StatsKeyMessage key : grpKeys.getValue())
+                keyGroups.put(key, grpKeys.getKey());
+
+        // NodeId to <Key to partitions on node> map
+        Map<UUID, Map<StatsKeyMessage, int[]>> reqMap = new HashMap<>();
+        for (Map.Entry<CacheGroupContext, Collection<StatsKeyMessage>> grpEntry : grpContexts.entrySet()) {
+            Map<UUID, int[]> reqNodes = nodePartitions(grpEntry.getKey(), null);
+            for (Map.Entry<UUID, int[]> nodeParts : reqNodes.entrySet())
+                reqMap.compute(nodeParts.getKey(), (k, v) -> {
+                    if (v == null)
+                        v = new HashMap<>();
+
+                    Collection<StatsKeyMessage> grpKeys = grpContexts.get(grpEntry.getKey());
+
+                    for (StatsKeyMessage key : grpKeys) {
+                        int[] keyNodeParts = (failedPartitions == null) ? nodeParts.getValue() :
+                                intersect(nodeParts.getValue(), failedPartitions.get(key));
+
+                        if (keyNodeParts.length > 0)
+                            v.put(key, keyNodeParts);
+                    }
+
+                    return v.size() > 0 ? v : null;
+                });
+        }
+
+        Collection<StatsCollectionAddrRequest> reqs = new ArrayList<>();
+        for (Map.Entry<UUID, Map<StatsKeyMessage, int[]>> nodeGpsParts: reqMap.entrySet()) {
+            if (nodeGpsParts.getValue().isEmpty())
+                continue;
+            StatsCollectionRequest req = prepareRequest(colId, nodeGpsParts.getValue());
+            reqs.add(new StatsCollectionAddrRequest(req, nodeGpsParts.getKey()));
+        }
+
+        return reqs;
+    }
+
+    /**
+     * Generate statistics collection requests by given keys.
+     *
+     * @param colId Collection id.
+     * @param keys Collection of keys to collect statistics by.
+     * @return Collection of statistics collection addressed request.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    protected Collection<StatsCollectionAddrRequest> generateCollectionRequests(
+        UUID colId,
+        Collection<StatsKeyMessage> keys,
+        Map<StatsKeyMessage, int[]> failedPartitions
+    ) throws IgniteCheckedException {
+        Map<CacheGroupContext, Collection<StatsKeyMessage>> grpContexts = extractGroups(keys);
+        return generateCollectionRequests(colId, keys, failedPartitions, grpContexts);
+    }
+
+    /**
+     * Intersect two specified arrays.
+     *
+     * @param a First array.
+     * @param b Second array.
+     * @return Arrays intersection.
+     */
+    protected static int[] intersect(int[] a, int[] b) {
+        if (a == null || b == null)
+            return new int[0];
+        Set<Integer> aSet = Arrays.stream(a).boxed().collect(Collectors.toSet());
```

Comment:
Can we be sure both arrays are sorted?
If so, intersect will of sorted arrays is trivial.
Or may be BitSet fits your needs better?

Reply:
Code removed.